### ISSUE TITLE
UnimplementedError: File system scheme '[local]' not implemented on TPU distribute strategy

### ISSUE THREAD

**Issue by h4ck4l1** [keras-team/keras-tuner#921]
When using kt.Hyperband i get this issue. I had checked all other forum answers including github issues. but they didn't work out as there's no concrete solution

Platform: Google Colab

colab notebook:
https://colab.research.google.com/drive/1q75FjE-3aX2ZtUpl3J7s1PTFEA2mL80a#scrollTo=leCrarx9OSd1

i use this
```python3

tpu_resolver = tf.distribute.cluster_resolver.TPUClusterResolver()
tf.config.experimental_connect_to_cluster(tpu_resolver)
tf.tpu.experimental.initialize_tpu_system(tpu_resolver)
strategy = tf.distribute.TPUStrategy(tpu_resolver)
tuner = kt.Hyperband(
    build_model,
    objective="val_accuracy",
    max_epochs=50,
    hyperband_iterations=3,
    factor=5,
    overwrite=True,
    distribution_strategy=strategy
)
tuner.search(train_ds,validation_data=test_ds,steps_per_epoch=steps,validation_steps=valid_step)
```
My error
```
Epoch 1/2
58/58 [==============================] - ETA: 0s - loss: 0.6915 - accuracy: 0.7893Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/base_tuner.py", line 270, in _try_run_and_update_trial
    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/base_tuner.py", line 235, in _run_and_update_trial
    results = self.run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/tuners/hyperband.py", line 425, in run_trial
    return super().run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/tuner.py", line 287, in run_trial
    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/tuner.py", line 214, in _build_and_fit_model
    results = self.hypermodel.fit(hp, model, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/hypermodel.py", line 144, in fit
    return model.fit(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/ops.py", line 1128, in _numpy
    raise core._status_to_exception(e) from None  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.UnimplementedError: File system scheme '[local]' not implemented (file: './untitled_project/trial_0002/checkpoint_temp/part-00000-of-00001')
	Encountered when executing an operation using EagerExecutor. This error cancels all future operations and poisons their output tensors.

```

When i try the solution somebody posted 
```python3

def new_on_epoch_end(self, epoch, logs=None):
    if not self.objective.has_value(logs):      
        save_locally = tf.saved_model.SaveOptions(experimental_io_device = '/job:localhost')
        self.model.save_weights(self.filepath, options = save_locally)
        return
    current_value = self.objective.get_value(logs)
    if self.objective.better_than(current_value, self.best_value):
        self.best_value = current_value
        save_locally = tf.saved_model.SaveOptions(experimental_io_device = '/job:localhost')
        self.model.save_weights(self.filepath, options = save_locally)
with patch('keras_tuner.engine.tuner_utils.SaveBestEpoch.on_epoch_end', new_on_epoch_end):
    tuner.search(train_ds,validation_data=test_ds,steps_per_epoch=steps,validation_steps=valid_step)
```

I get this error:

```
Epoch 1/2
58/58 [==============================] - ETA: 0s - loss: 1.2381 - accuracy: 0.6274Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/base_tuner.py", line 270, in _try_run_and_update_trial
    self._run_and_update_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/base_tuner.py", line 235, in _run_and_update_trial
    results = self.run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/tuners/hyperband.py", line 425, in run_trial
    return super().run_trial(trial, *fit_args, **fit_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/tuner.py", line 287, in run_trial
    obj_value = self._build_and_fit_model(trial, *args, **copied_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/tuner.py", line 214, in _build_and_fit_model
    results = self.hypermodel.fit(hp, model, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras_tuner/engine/hypermodel.py", line 144, in fit
    return model.fit(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "<ipython-input-8-ec3edafd4088>", line 10, in new_on_epoch_end
    self.model.save_weights(self.filepath, options = save_locally)
AttributeError: 'SaveOptions' object has no attribute 'experimental_enable_async_checkpoint'

```

please help!

**Comment by aaarrti** [comment]
Afaik, colab TPUs are using TPU-node architecture, which means they are running on a different host, than the VM, which serves colab runtime, so TPU can't read the local file-system of colab VM (seems logical). One workaround you can do is: 
- create google cloud project 
- Add this one top of your notebook
```python

!gcloud auth login
!gcloud config set project [PROJECT-id]

from google.colab import auth
auth.authenticate_user()

```
- Create GCS bucket
```terraform
resource "google_storage_bucket" "data_bucket" {
  name                        = var.bucket-name
  location                    = var.region
  storage_class               = "STANDARD"
  uniform_bucket_level_access = true
  force_destroy               = true
}
```
- Then you initialize TPU cluster following way
```python
tf.distribute.cluster_resolver.TPUClusterResolver(tpu='grpc://' + os.environ['COLAB_TPU_ADDR'])
```
- At this point, you can try to load dataset from your bucket, and you will get a similar error

```
{
  "code": 403,
  "message": "service-<id>@cloud-tpu.iam.gserviceaccount.com does not have storage.objects.list access to the Google Cloud Storage bucket. Permission 'storage.objects.list' denied on resource (or it may not exist).",
}
```
- Then you go, and add IAM role bindings for TPUs service account
```terraform
resource "google_storage_bucket_iam_member" "TPU_ROLE_1" {
  bucket     = var.bucket-name
  role       = "roles/storage.objectViewer"
  member     = var.tpu_sa
  depends_on = [google_storage_bucket.data_bucket]
}

resource "google_storage_bucket_iam_member" "TPU_ROLE_2" {
  bucket     = var.bucket-name
  role       = "roles/storage.legacyBucketReader"
  member     = var.tpu_sa
  depends_on = [google_storage_bucket.data_bucket]
}

resource "google_storage_bucket_iam_member" "TPU_ROLE_3" {
  bucket     = var.bucket-name
  role       = "roles/storage.objectCreator"
  member     = var.tpu_sa
  depends_on = [google_storage_bucket.data_bucket]
}

resource "google_storage_bucket_iam_member" "TPU_ROLE_4" {
  bucket     = var.bucket-name
  role       = "roles/storage.legacyBucketWriter"
  member     = var.tpu_sa
  depends_on = [google_storage_bucket.data_bucket]
}
```
where `var.tpu_sa` is `"serviceAccount:service-<id>@cloud-tpu.iam.gserviceaccount.com"`
- Then finally, your colab TPU should be able to access GCS bucket.

**Alternatively**, you could use [Kaggle Notebooks](https://www.kaggle.com/discussions/product-feedback/369338), since they provide TPU VMs, so you can read the local file-system when running those.

To sum up, this is not a `keras-tuner` bug 🙈 , but rather an environment misconfiguration.
